### PR TITLE
Add input and output traits

### PR DESCRIPTION
**Pull request by mtdowling**
These traits allow shapes to be specialized as only acting as input or
output for a single operation. This will make it easier to generate code
from Smithy models, and we will be able to give more relaxed backward
compatibility semantics to structures that are marked as `@input`
structures.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by JordonPhillips — CHANGES_REQUESTED** [review]

File: smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/OperationInputOutputValidator.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    private ValidationEvent emitInvalidMemberRef(MemberShape member, String trait) {
+        return error(member, "This member illegally references a structure marked with the @"
+                             + trait + " trait: " + member.getTarget());
+    }
+
+    private ValidationEvent emitBadInputOutputName(Shape operation, String property, ShapeId target) {
+        return error(operation, String.format(
+                        "The %s of this operation should target a shape that starts with the operation's name, '%s', "
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/OperationInputOutputValidator.java:98]: I don't think we need to call out inline syntax here. This is a validation event dealing with the semantic model, so I don't think we should call out specific model representations in the validation.

File: smithy-model/src/main/resources/software/amazon/smithy/model/loader/prelude.smithy [hunk 2]
```diff
@@ -817,3 +817,13 @@ structure httpChecksumProperty {
     },
 ])
 string HttpChecksumLocation
+
+/// Specializes a structure for use only as the input of a single operation.
+@trait(selector: "structure", conflicts: [output, error])
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/main/resources/software/amazon/smithy/model/loader/prelude.smithy:822]: Yes, it works with mixins. Every trait works with mixins. I can add a test case.

File: smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/OperationInputOutputValidator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (JordonPhillips) [reply smithy-model/src/main/java/software/amazon/smithy/model/validation/validators/OperationInputOutputValidator.java:98]: I disagree. Essentially everyone will be writing models in the IDL format, so giving them guidance on things that can make things easier there will always be better than not. We can preface that with "In the IDL," or something to make it clear when the advice applies though.

File: docs/source/1.0/spec/core/type-refinement-traits.rst [hunk 3]
```diff
@@ -138,6 +138,62 @@ in Java).
...
+Conflicts with
+    :ref:`input-trait`, :ref:`error-trait`
+
+Structure shapes marked with the ``@input`` trait MUST adhere to the
+following constraints:
+
+1. They can only be referenced in the model as an operation's input.
+2. They cannot be used as the input of more than one operation.
+3. They MUST have a shape name that starts with the name of any operation
+   they are associated with.
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply docs/source/1.0/spec/core/type-refinement-traits.rst:163]: Good points. I updated this to be a WARNING instead.

**Review by JordonPhillips — APPROVED** [review]

**Comment by mtdowling** [comment]
Closing in favor of #980

**Pull request closed without merging**